pkg/deploy/deployer.go

```diff
@@ -0,0 +1,311 @@
+	// if RP_PREDEPLOY is not set, we read old rpServicePrincipalID and return it
+	if len(os.Getenv("RP_PREDEPLOY")) != 0 {
+		var data []byte
+		data, err = generator.GenerateNSGTemplate(true)
+		if err != nil {
+			return
+		}
+
+		var azuretemplate map[string]interface{}
+		err = json.Unmarshal(data, &azuretemplate)
+		if err != nil {
+			return
+		}
+
+		log.Infof("deploying nsg and managedIdentity to %s", d.resourceGroupName)
+		err = d.deployments.CreateOrUpdateAndWait(ctx, d.resourceGroupName, "rp-production-nsg", azresources.Deployment{
+			Properties: &azresources.DeploymentProperties{
+				Template: azuretemplate,
+				Mode:     azresources.Incremental,
+			},
+		})
+		if err != nil {
+			return
+		}
+	}
+	deployment, err := d.deployments.Get(ctx, d.resourceGroupName, "rp-production-nsg")
+	if err != nil {
+		return
+	}
+	return deployment.Properties.Outputs.(map[string]interface{})["rpServicePrincipalId"].(map[string]interface{})["value"].(string), nil
+}
+
+func (d *deployer) Deploy(ctx context.Context, log *logrus.Entry, rpServicePrincipalID string) error {
+	data, err := generator.GenerateRPTemplate(true)
+	if err != nil {
+		return err
+	}
+
+	var azuretemplate map[string]interface{}
+	err = json.Unmarshal(data, &azuretemplate)
+	if err != nil {
+		return err
+	}
+
+	parameters, err := getParameters()
+	if err != nil {
+		return err
+	}
+	parameters["vmssName"].(map[string]interface{})["value"] = d.version
+	parameters["rpServicePrincipalId"].(map[string]interface{})["value"] = rpServicePrincipalID
+	// azure enforce ^[a-z][a-z0-9-]{1,61}[a-z0-9]$. for DomainNameLabel.
+	// gitCommit do not comply as it might start with a number
+	parameters["vmssDomainNameLabel"].(map[string]interface{})["value"] = "vmss-" + d.version
+
+	log.Infof("deploying rp version %s to %s", d.version, d.resourceGroupName)
+	return d.deployments.CreateOrUpdateAndWait(ctx, d.resourceGroupName, "rp-production-"+d.version, azresources.Deployment{
+		Properties: &azresources.DeploymentProperties{
+			Template:   azuretemplate,
+			Mode:       azresources.Incremental,
+			Parameters: parameters,
+		},
+	})
+}
+
+func (d *deployer) Upgrade(ctx context.Context, log *logrus.Entry) error {
+	scaleSets, err := d.listVMSS(ctx)
+	if err != nil {
+		return err
+	}
+
+	{
+		log.Info("checking new RP health")
+		for _, vmss := range scaleSets {
+			if *vmss.Name == "rp-vmss-"+d.version {
+				log.Info("check VMSS " + *vmss.Name)
+				scaleSetsVMs, err := d.listVMSSVMs(ctx, *vmss.Name)
+				if err != nil {
+					return err
+				}
+
+				// construct readiness tracking map with FQDN for checking
+				readyMap := make(map[string]bool, len(scaleSetsVMs))
+				for _, vm := range scaleSetsVMs {
+					pi, err := d.network.GetVirtualMachineScaleSetPublicIPAddress(ctx, d.resourceGroupName, *vmss.Name, *vm.InstanceID, "rp-vmss-nic", "rp-vmss-ipconfig", "rp-vmss-pip", "")
+					if err != nil {
+						return err
+					}
```

Comment:
you can simplify: the VMSS VMs will be accessible at vm{0,1,...}.$vmssDomainNameLabel.$location.cloudapp.azure.com.